### PR TITLE
Add `group_size` arg to decorrelated batch normalization

### PR DESCRIPTION
**Pull request by toslunar**
Part of #7905.

### CODE REVIEW

**Comment by hvy** [comment]
This is already a "no compat" as it's changing the default behavior?

**Comment by toslunar** [comment]
> This is already a "no compat" as it's changing the default behavior?

Really?  I suppose the function node (`DecorrelatedBatchNormalization`) is internal.

**Comment by hvy** [comment]
Oh you are right, I was only looking at the function node.

Another question is how beneficial it is to introduce another similar argument. The reasons for asking the question is that
- it's trivial to compute the number of groups for the users
- having two arguments for similar purposes is a bit confusing. Of course, we could deprecate/remove `groups` in the future
- We have similar arguments in e.g. group normalization (`groups`) so users should be familiar

I agree that the default behavior must be changed. Another option would be to simply change the default behavior, leaving the interface unchanged?

**Comment by toslunar** [comment]
> * it's trivial to compute the number of groups for the users

It should not burden users to control size of each group.

> * having two arguments for similar purposes is a bit confusing.

Do you request something like `groups={'size': 16}`?

> Of course, we could deprecate/remove `groups` in the future

I don't want to remove `groups`.

> * We have similar arguments in e.g. group normalization (`groups`) so users should be familiar

Users are familiar with `in_size`, `out_size`, and `ksize`.

**Comment by hvy** [comment]
I see, so you want to keep both and would also have implemented it that way if you'd be to write this function/link from scratch? We have this `groups` argument in group normalization, convolution, deconvolution as well where it has the same meaning (splitting along the channel axis) which I think is consistent and thus alleviating the burden. I personally find introducing another argument to DBN because the default value differed from the value in the paper isn't really justifying this change. But maybe I'm missing your point.

I also tried responding to you responses above. Sorry about the length.

> it's trivial to compute the number of groups for the users
>> It should not burden users to control size of each group.

I think this is a tradeoff. Arguments that shadow each other or lead to errors being inconsistent isn't necessarily introducing less burden.

> having two arguments for similar purposes is a bit confusing.
>> Do you request something like groups={'size': 16}?

No, I meant keeping the current type and semantics of `groups`.

> Of course, we could deprecate/remove groups in the future
>> I don't want to remove groups.

This depends I think.

> We have similar arguments in e.g. group normalization (groups) so users should be familiar
>> Users are familiar with in_size, out_size, and ksize.

You're right. But my point was that the current interface isn't foreign neither. However, adding another `*_size` argument might make it so.

**Comment by toslunar** [comment]
My opinions are the following. (Not necessarily summarizing the authors' opinions.)

- Section 3.4 of the paper says that `k_G` is the hyperparameter to be controlled for limiting computational complexity because grouped decorrelated BN is basically `k_G`-times slower than BN is.
- Section 4.1 (MLP example) has a suggestion that `k_G` has to be much smaller than batchsize from the poor result for `k_G = 128 < 256 = batchsize`.
  - But we rarely have to care about it if `batchsize := N * W * H`.
